src/systems/elevator/Elevator.cc

```diff
@@ -0,0 +1,449 @@
+  /// \brief Updates the elevator state based on the current cabin position and
+  /// then publishes the new state
+  /// \param[in] _info Current simulation step info
+  /// \param[in] _ecm Entity component manager
+  public: void UpdateState(const ignition::gazebo::UpdateInfo &_info,
+                           const EntityComponentManager &_ecm);
+
+  /// \brief Callback for the door lidar scans
+  /// \param[in] _floorLevel Floor level
+  /// \param[in] _msg Laserscan message
+  public: void OnLidarMsg(size_t _floorLevel, const msgs::LaserScan &_msg);
+
+  /// \brief Callback for the elevator commands
+  /// \param[in] _msg Message that carries the floor level target
+  public: void OnCmdMsg(const msgs::Int32 &_msg);
+
+  /// \brief Ignition communication node
+  public: transport::Node node;
+
+  /// \brief Model to which this system belongs
+  public: Model model;
+
+  /// \brief Joints of the doors of the elevator
+  public: std::vector<Entity> doorJoints;
+
+  /// \brief Joint of the cabin
+  public: Entity cabinJoint;
+
+  /// \brief State vector that identifies whether the doorway on each floor
+  /// level is blocked
+  public: std::vector<bool> isDoorwayBlockedStates;
+
+  /// \brief Timer that keeps the door at the target floor level open
+  public: DoorTimer doorTimer;
+
+  /// \brief Monitor that checks whether the door at the target floor level has
+  /// been opened or closed
+  public: JointMonitor doorJointMonitor;
+
+  /// \brief Monitor that checks whether the cabin has reached the target floor
+  /// level
+  public: JointMonitor cabinJointMonitor;
+
+  /// \brief System update period calculated from <update_rate>
+  public: std::chrono::steady_clock::duration updatePeriod{0};
+
+  /// \brief Last system update simulation time
+  public: std::chrono::steady_clock::duration lastUpdateTime{0};
+
+  /// \brief State publish period calculated from <state_publish_rate>
+  public: std::chrono::steady_clock::duration statePubPeriod{0};
+
+  /// \brief Last state publish simulation time
+  public: std::chrono::steady_clock::duration lastStatePubTime{0};
+
+  /// \brief Elevator state publisher
+  public: transport::Node::Publisher statePub;
+
+  /// \brief Elevator state message
+  public: msgs::Int32 stateMsg;
+
+  /// \brief Elevator state machine
+  public: std::unique_ptr<ElevatorStateMachine> stateMachine;
+};
+
+//////////////////////////////////////////////////
+Elevator::Elevator() : dataPtr(std::make_shared<ElevatorPrivate>()) {}
+
+//////////////////////////////////////////////////
+void Elevator::Configure(const Entity &_entity,
+                         const std::shared_ptr<const sdf::Element> &_sdf,
+                         EntityComponentManager &_ecm,
+                         EventManager & /*_eventMgr*/)
+{
+  this->dataPtr->model = Model(_entity);
+
+  // Initialize system update period
+  double rate = _sdf->Get<double>("update_rate", 10).first;
+  std::chrono::duration<double> period{rate > 0 ? 1 / rate : 0};
```

Comment:
Should we just print an error if the period is <= 0 and stop loading the plugin, or does the plugin have any functionality with zero?

Reply:
Zero means no throttling. I'll add it to the documentation (or treat it as an error as you suggested)